### PR TITLE
[SC-190] Apply synapse user profile info as tags

### PR DESCRIPTION
**Pull request by zaro0508**
Refactor to get synapse user profile info[1] and apply that info as
tags on service catalog resources.  The tag keys will be  prepended with
'synapse:'.

[1] https://docs.synapse.org/rest/org/sagebionetworks/repo/model/UserProfile.html

### CODE REVIEW

File: set_bucket_tags/app.py [hunk 1]
```diff
@@ -54,27 +53,49 @@ def get_principal_id(tags):
...
+def merge(list_x, list_y):
+  '''Merge two list of tags'''
+  z = []
+  for x in list_x:
+    z.append(x)
+
+  for y in list_y:
+    z.append(y)
+
+  return z
```
Comment (tthyer) [inline set_bucket_tags/app.py:96]: you can shorten this merge function to `list_x + list_y`

File: set_bucket_tags/app.py [hunk 2]
```diff
@@ -83,15 +104,18 @@ def create_or_update(event, context):
...
-  tags = get_bucket_tags(bucket_name)
-  principal_id = get_principal_id(tags)
-  email = get_synapse_email(principal_id)
-  tags = add_owner_email_tag(tags, email)
+  bucket_tags = get_bucket_tags(bucket_name)
+  principal_id = get_principal_id(bucket_tags)
+  user_profile = get_synapse_userProfile(principal_id)
+  synapse_tags = get_synapse_tags(user_profile)
+  # put_bucket_tagging is a replace operation.  need to give it all
+  # tags otherwise it will remove existing tags not in the list
```
Comment (tthyer) [inline set_bucket_tags/app.py:112]: That comment is a good idea

File: set_instance_tags/app.py [hunk 3]
```diff
@@ -57,43 +56,35 @@ def get_principal_id(tags):
...
+  for key, value in user_profile.items():
+    if key in IGNORE_KEYS:
+      continue
 
-  f_tags = []
+    # derive synapse email tag based on userName
+    if key == "userName":
+      synapse_email = f'{value}@synapse.org'
+      tags.append({'Key': 'synapse:email', 'Value': synapse_email})
+      tags.append({'Key': 'OwnerEmail', 'Value': synapse_email})  # legacy
```
Comment (tthyer) [inline set_instance_tags/app.py:81]: Nice, I like using the `synapse:` prefix

**Review by tthyer — CHANGES_REQUESTED** [review]
Looks good, I only recommend that you don't need the merge function, you can just add the lists

File: set_bucket_tags/app.py (hunk 1 above)
Comment (zaro0508) [reply set_bucket_tags/app.py]: thanks.  fixed.

File: tests/unit/set_bucket_tags/test_get_synapse_tags.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+from set_instance_tags import app
+
+jsmith_profile = {
+  "createdOn": "2020-06-18T16:34:18.000Z",
+  "firstName": "Joe",
+  "lastName": "Smith",
+  "ownerId": "1111111",
+  "userName": "jsmith"
+}
```
Comment (cascadianblue) [inline tests/unit/set_bucket_tags/test_get_synapse_tags.py:11]: Constants should be all uppercase

File: tests/unit/set_bucket_tags/test_get_synapse_userProfile.py [hunk 5]
```diff
@@ -5,14 +5,15 @@
...
 from set_instance_tags import app
 
+jsmith_profile = {
+  "createdOn": "2020-06-18T16:34:18.000Z",
+  "firstName": "Joe",
+  "lastName": "Smith",
+  "ownerId": "1111111",
+  "userName": "jsmith"
+}
+
```
Comment (cascadianblue) [inline tests/unit/set_bucket_tags/test_get_synapse_userProfile.py:15]: Constants should be all uppercase

**Review by cascadianblue — CHANGES_REQUESTED** [review]

**Comment by cascadianblue** [comment]
This sentence in the README needs correction:

> Cloudformation Custom Resource that sets tags Synapse tags on AWS resources.

**Review by tthyer — APPROVED** [review]

File: set_instance_tags/app.py [hunk 6]
```diff
@@ -57,43 +56,35 @@ def get_principal_id(tags):
...
+def get_synapse_userProfile(synapse_id):
+  '''Get synapse user profile data'''
   syn = synapseclient.Synapse()
   user_profile = syn.getUserProfile(synapse_id)
   log.debug(f'Synapse user profile: {user_profile}')
-  user_name = user_profile.get('userName')
-  synapse_email = f'{user_name}@synapse.org'
-  return synapse_email
 
+  return user_profile
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply set_instance_tags/app.py]: fixed

File: tests/unit/set_bucket_tags/test_get_synapse_userProfile.py (hunk 5 above)
Comment (zaro0508) [reply tests/unit/set_bucket_tags/test_get_synapse_userProfile.py]: fixed

File: tests/unit/set_bucket_tags/test_get_synapse_tags.py (hunk 4 above)
Comment (zaro0508) [reply tests/unit/set_bucket_tags/test_get_synapse_tags.py]: fixed

File: set_bucket_tags/app.py [hunk 7]
```diff
@@ -54,24 +53,34 @@ def get_principal_id(tags):
...
+    if key in IGNORE_KEYS:
+      continue
+
+    # derive synapse email tag based on userName
+    if key == "userName":
+      synapse_email = f'{value}@synapse.org'
+      tags.append({'Key': 'synapse:email', 'Value': synapse_email})
+      tags.append({'Key': 'OwnerEmail', 'Value': synapse_email})  # legacy
+
+    tag = {'Key': f'synapse:{key}', 'Value': value}
```
Comment (ahayden) [inline set_bucket_tags/app.py:80]: Remember that any tag that that doesn't start with `Protected/` can be overwritten by any other instance.  IMO these should all be under the `Protected/` tag path *or* we remove the policy for instances to apply tags to themselves (is there another PR that does that?)

**Review by ahayden — APPROVED** [review]
This should be a lot more resilient than those scripts

**Review by ahayden — CHANGES_REQUESTED** [review]
Changing status to requesting changes because these tags should be under the `Protected/` path

**Comment by zaro0508** [comment]
> Changing status to requesting changes because these tags should be under the `Protected/` path

We have a messed up policy if an EC2 can over write another EC2's tags.  Let's solve that problem independently from this PR.  created an issue for it here: https://sagebionetworks.jira.com/browse/SC-241

**Review by brucehoff — APPROVED** [review]

**Review by ahayden — APPROVED** [review]

**Comment by ahayden** [comment]
We'll update the tagging policies to disallow cross-EC2 tagging

**Pull request merged**